src/main/java/com/enderio/base/common/loot/providers/nbt/ImprovedContextNbtProvider.java

```diff
@@ -0,0 +1,141 @@
+import com.google.common.collect.ImmutableSet;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+import java.util.Set;
+import javax.annotation.Nullable;
+import net.minecraft.advancements.critereon.NbtPredicate;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.util.GsonHelper;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.storage.loot.GsonAdapterFactory;
+import net.minecraft.world.level.storage.loot.LootContext;
+import net.minecraft.world.level.storage.loot.parameters.LootContextParam;
+import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
+import net.minecraft.world.level.storage.loot.providers.nbt.LootNbtProviderType;
+import net.minecraft.world.level.storage.loot.providers.nbt.NbtProvider;
+
+/**
+ * A NbtProvider that provides either the
+ * {@linkplain LootContextParams#BLOCK_ENTITY block entity}'s NBT data or an
+ * entity's NBT data based on an {@link LootContext.EntityTarget} but without
+ * the id and positional data.
+ */
+public class ImprovedContextNbtProvider implements NbtProvider {
+    private static final String BLOCK_ENTITY_ID = "block_entity";
+
+    private static final ImprovedContextNbtProvider.Getter BLOCK_ENTITY_PROVIDER = new ImprovedContextNbtProvider.Getter() {
+        public Tag get(LootContext p_165582_) {
+            BlockEntity blockentity = p_165582_.getParamOrNull(LootContextParams.BLOCK_ENTITY);
+            if (blockentity == null) {
+                return null;
+            }
+            CompoundTag tag = blockentity.saveWithoutMetadata();
+            return tag.isEmpty() ? null : tag;
+        }
+
+        public String getId() {
+            return BLOCK_ENTITY_ID;
+        }
+
+        public Set<LootContextParam<?>> getReferencedContextParams() {
+            return ImmutableSet.of(LootContextParams.BLOCK_ENTITY);
+        }
+    };
+    public static final ImprovedContextNbtProvider BLOCK_ENTITY = new ImprovedContextNbtProvider(BLOCK_ENTITY_PROVIDER);
+    final ImprovedContextNbtProvider.Getter getter;
+
+    private static ImprovedContextNbtProvider.Getter forEntity(final LootContext.EntityTarget pEntityTarget) {
+        return new ImprovedContextNbtProvider.Getter() {
+            @Nullable
+            public Tag get(LootContext p_165589_) {
+                Entity entity = p_165589_.getParamOrNull(pEntityTarget.getParam());
+                return entity != null ? NbtPredicate.getEntityTagToCompare(entity) : null;
+            }
+
+            public String getId() {
+                return pEntityTarget.getName();
+            }
+
+            public Set<LootContextParam<?>> getReferencedContextParams() {
+                return ImmutableSet.of(pEntityTarget.getParam());
```

Comment:
use Set.of

Reply:
The `ContextNbtProvider` that I shamelessly copied this from uses this, so I'm reluctant to change it, but if you think it's okay, then I will.

Reply:
it's effectivly the same, and at that point it's personal preference. ImmutableSet could indicate, that someone wants implementationdetails for a specific reason, while the set just indicates, that you want a set